### ISSUE TITLE
Can't compile with build-bindgen

### ISSUE THREAD

**Issue by jiayihu** [Veykril/wasm3-rs#20]
I can't get to compile to crate in any way. I've installed clang from brew and also MacOS Xcode devtools, but if I use the `build-bindgen` features I get the following error:

```
Caused by:
  process didn't exit successfully: `/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/debug/build/wasm3-sys-c75d8740be89c22d/build-script-build` (exit code: 101)
  --- stderr
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/machine/_types.h:34:2: error: architecture not supported
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:55:9: error: unknown type name '__int64_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:56:9: error: unknown type name '__int32_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:57:9: error: unknown type name '__int32_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:60:9: error: unknown type name '__uint32_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:61:9: error: unknown type name '__uint32_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:62:9: error: unknown type name '__uint64_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:68:9: error: unknown type name '__darwin_natural_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:70:9: error: unknown type name '__uint16_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:71:9: error: unknown type name '__int64_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:72:9: error: unknown type name '__int32_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:73:9: error: unknown type name '__uint32_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:74:9: error: unknown type name '__int32_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:75:9: error: unknown type name '__uint32_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:76:9: error: unknown type name '__uint32_t'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/machine/types.h:37:2: error: architecture not supported
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types/_va_list.h:32:9: error: unknown type name '__darwin_va_list'
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types/_size_t.h:31:9: error: unknown type name '__darwin_size_t'
  fatal error: too many errors emitted, stopping now [-ferror-limit=]
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/cdefs.h:807:2: error: Unsupported architecture, err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/machine/_types.h:34:2: error: architecture not supported, err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:55:9: error: unknown type name '__int64_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:56:9: error: unknown type name '__int32_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:57:9: error: unknown type name '__int32_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:60:9: error: unknown type name '__uint32_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:61:9: error: unknown type name '__uint32_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:62:9: error: unknown type name '__uint64_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:68:9: error: unknown type name '__darwin_natural_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:70:9: error: unknown type name '__uint16_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:71:9: error: unknown type name '__int64_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:72:9: error: unknown type name '__int32_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:73:9: error: unknown type name '__uint32_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:74:9: error: unknown type name '__int32_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:75:9: error: unknown type name '__uint32_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types.h:76:9: error: unknown type name '__uint32_t', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/machine/types.h:37:2: error: architecture not supported, err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types/_va_list.h:32:9: error: unknown type name '__darwin_va_list', err: true
  /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.15.sdk/usr/include/sys/_types/_size_t.h:31:9: error: unknown type name '__darwin_size_t', err: true
  fatal error: too many errors emitted, stopping now [-ferror-limit=], err: true
  thread 'main' panicked at 'Failed to generate bindings: ()', /Users/jiayihu/.cargo/registry/src/github.com-1ecc6299db9ec823/wasm3-sys-0.1.2/build.rs:121:10
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Trying instead to compile it without the bindgen feature results in the following error:

```
error: failed to run custom build command for `wasm3-sys v0.1.2`

Caused by:
  process didn't exit successfully: `/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/debug/build/wasm3-sys-ed19931fe417e3c8/build-script-build` (exit code: 101)
  --- stderr
  wasm3/source/m3_core.h:11:10: fatal error: 'stdio.h' file not found
  wasm3/source/m3_core.h:11:10: fatal error: 'stdio.h' file not found, err: true
  thread 'main' panicked at 'Unable to generate bindings: ()', /Users/jiayihu/.cargo/registry/src/github.com-1ecc6299db9ec823/bindgen-0.55.1/src/main.rs:54:36
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
  thread 'main' panicked at 'Failed to run bindgen: ExitStatus(ExitStatus(256))', /Users/jiayihu/.cargo/registry/src/github.com-1ecc6299db9ec823/wasm3-sys-0.1.2/build.rs:80:9
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I've also tried to just run `clang -v a.c` where `a.c` contains only an include of stdio.h and it works without issue, meaning it's able to find the header file. So the issue is in rust bindgen or something in this crate I suppose.

**Comment by Veykril** [comment]
Are you certain that you properly set up clang? I unfortunately do not have access to a MacOS machine so I can't test this and the stderr output doesn't really help me out either. But I do believe this problem lies with either you clang setup or bindgen itself from looking through its issue tracker.

**Comment by jiayihu** [comment]
I think I resolved the issue because I'm cross compiling to `thumbv7em-none-eabi` as target. By running

```
BINDGEN_EXTRA_CLANG_ARGS="--sysroot=/usr/local/opt/gcc-arm-none-eabi/arm-none-eabi" cargo build
```

Seems to work, although another issue appears.

```
The following warnings were emitted during compilation:

warning: wasm3/source/m3_api_wasi.c: In function 'convert_clockid':
warning: wasm3/source/m3_api_wasi.c:144:52: error: 'CLOCK_MONOTONIC' undeclared (first use in this function); did you mean '__WASI_CLOCK_MONOTONIC'?
warning:      case __WASI_CLOCK_MONOTONIC:            return CLOCK_MONOTONIC;
warning:                                                     ^~~~~~~~~~~~~~~
warning:                                                     __WASI_CLOCK_MONOTONIC
warning: wasm3/source/m3_api_wasi.c:144:52: note: each undeclared identifier is reported only once for each function it appears in
warning: wasm3/source/m3_api_wasi.c:145:52: error: 'CLOCK_PROCESS_CPUTIME_ID' undeclared (first use in this function); did you mean '__WASI_CLOCK_PROCESS_CPUTIME_ID'?
warning:      case __WASI_CLOCK_PROCESS_CPUTIME_ID:   return CLOCK_PROCESS_CPUTIME_ID;
warning:                                                     ^~~~~~~~~~~~~~~~~~~~~~~~
warning:                                                     __WASI_CLOCK_PROCESS_CPUTIME_ID
warning: wasm3/source/m3_api_wasi.c:147:52: error: 'CLOCK_THREAD_CPUTIME_ID' undeclared (first use in this function); did you mean 'CLOCK_PROCESS_CPUTIME_ID'?
warning:      case __WASI_CLOCK_THREAD_CPUTIME_ID:    return CLOCK_THREAD_CPUTIME_ID;
warning:                                                     ^~~~~~~~~~~~~~~~~~~~~~~
warning:                                                     CLOCK_PROCESS_CPUTIME_ID
warning: In file included from wasm3/source/m3_api_wasi.c:12:0:
warning: wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_fd_fdstat_get':
warning: wasm3/source/m3_api_wasi.c:310:31: error: 'O_DSYNC' undeclared (first use in this function); did you mean 'O_SYNC'?
warning:                         ((fl & O_DSYNC)     ? __WASI_FDFLAG_DSYNC     : 0) |
warning:                                ^
warning: wasm3/source/m3_api_defs.h:37:58: note: in definition of macro 'm3ApiWriteMem16'
warning:  #  define m3ApiWriteMem16(ptr, val)  { * (u16 *)(ptr) = (val); }
warning:                                                           ^~~
warning: wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_fd_seek':
warning: wasm3/source/m3_api_wasi.c:353:11: warning: implicit declaration of function 'lseek'; did you mean 'fseek'? [-Wimplicit-function-declaration]
warning:      ret = lseek(fd, offset, whence);
warning:            ^~~~~
warning:            fseek
warning: wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_path_open':
warning: wasm3/source/m3_api_wasi.c:422:58: error: 'O_DSYNC' undeclared (first use in this function); did you mean 'O_SYNC'?
warning:                  ((fs_flags & __WASI_FDFLAG_DSYNC)      ? O_DSYNC     : 0) |
warning:                                                           ^~~~~~~
warning:                                                           O_SYNC
warning: wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_fd_read':
warning: wasm3/source/m3_api_wasi.c:474:19: warning: implicit declaration of function 'read'; did you mean 'fread'? [-Wimplicit-function-declaration]
warning:          int ret = read (fd, addr, len);
warning:                    ^~~~
warning:                    fread
warning: wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_fd_write':
warning: wasm3/source/m3_api_wasi.c:509:19: warning: implicit declaration of function 'write'; did you mean 'fwrite'? [-Wimplicit-function-declaration]
warning:          int ret = write (fd, addr, len);
warning:                    ^~~~~
warning:                    fwrite
warning: wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_fd_close':
warning: wasm3/source/m3_api_wasi.c:524:15: warning: implicit declaration of function 'close'; did you mean 'pclose'? [-Wimplicit-function-declaration]
warning:      int ret = close(fd);
warning:                ^~~~~
warning:                pclose
warning: wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_clock_res_get':
warning: wasm3/source/m3_api_wasi.c:596:9: warning: implicit declaration of function 'clock_getres' [-Wimplicit-function-declaration]
warning:      if (clock_getres(clk, &tp) != 0) {
warning:          ^~~~~~~~~~~~
warning: wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_clock_time_get':
warning: wasm3/source/m3_api_wasi.c:618:9: warning: implicit declaration of function 'clock_gettime'; did you mean 'localtime'? [-Wimplicit-function-declaration]
warning:      if (clock_gettime(clk, &tp) != 0) {
warning:          ^~~~~~~~~~~~~
warning:          localtime

error: failed to run custom build command for `wasm3-sys v0.1.2`

Caused by:
  process didn't exit successfully: `/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/debug/build/wasm3-sys-ed19931fe417e3c8/build-script-build` (exit code: 1)
  --- stdout
  TARGET = Some("thumbv7em-none-eabi")
  OPT_LEVEL = Some("0")
  HOST = Some("x86_64-apple-darwin")
  CC_thumbv7em-none-eabi = None
  CC_thumbv7em_none_eabi = None
  TARGET_CC = None
  CC = None
  CROSS_COMPILE = None
  CFLAGS_thumbv7em-none-eabi = None
  CFLAGS_thumbv7em_none_eabi = None
  TARGET_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  CARGO_CFG_TARGET_FEATURE = Some("dsp,mclass,thumb-mode,v5te,v6,v6k,v6t2,v7")
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_optimize.o" "-c" "wasm3/source/m3_optimize.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_compile.o" "-c" "wasm3/source/m3_compile.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_api_libc.o" "-c" "wasm3/source/m3_api_libc.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_env.o" "-c" "wasm3/source/m3_env.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_exec.o" "-c" "wasm3/source/m3_exec.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_bind.o" "-c" "wasm3/source/m3_bind.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_info.o" "-c" "wasm3/source/m3_info.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_emit.o" "-c" "wasm3/source/m3_emit.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_parse.o" "-c" "wasm3/source/m3_parse.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_module.o" "-c" "wasm3/source/m3_module.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_api_tracer.o" "-c" "wasm3/source/m3_api_tracer.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_core.o" "-c" "wasm3/source/m3_core.c"
  exit code: 0
  running: "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_api_wasi.o" "-c" "wasm3/source/m3_api_wasi.c"
  cargo:warning=wasm3/source/m3_api_wasi.c: In function 'convert_clockid':
  cargo:warning=wasm3/source/m3_api_wasi.c:144:52: error: 'CLOCK_MONOTONIC' undeclared (first use in this function); did you mean '__WASI_CLOCK_MONOTONIC'?
  cargo:warning=     case __WASI_CLOCK_MONOTONIC:            return CLOCK_MONOTONIC;
  cargo:warning=                                                    ^~~~~~~~~~~~~~~
  cargo:warning=                                                    __WASI_CLOCK_MONOTONIC
  cargo:warning=wasm3/source/m3_api_wasi.c:144:52: note: each undeclared identifier is reported only once for each function it appears in
  cargo:warning=wasm3/source/m3_api_wasi.c:145:52: error: 'CLOCK_PROCESS_CPUTIME_ID' undeclared (first use in this function); did you mean '__WASI_CLOCK_PROCESS_CPUTIME_ID'?
  cargo:warning=     case __WASI_CLOCK_PROCESS_CPUTIME_ID:   return CLOCK_PROCESS_CPUTIME_ID;
  cargo:warning=                                                    ^~~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=                                                    __WASI_CLOCK_PROCESS_CPUTIME_ID
  cargo:warning=wasm3/source/m3_api_wasi.c:147:52: error: 'CLOCK_THREAD_CPUTIME_ID' undeclared (first use in this function); did you mean 'CLOCK_PROCESS_CPUTIME_ID'?
  cargo:warning=     case __WASI_CLOCK_THREAD_CPUTIME_ID:    return CLOCK_THREAD_CPUTIME_ID;
  cargo:warning=                                                    ^~~~~~~~~~~~~~~~~~~~~~~
  cargo:warning=                                                    CLOCK_PROCESS_CPUTIME_ID
  cargo:warning=In file included from wasm3/source/m3_api_wasi.c:12:0:
  cargo:warning=wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_fd_fdstat_get':
  cargo:warning=wasm3/source/m3_api_wasi.c:310:31: error: 'O_DSYNC' undeclared (first use in this function); did you mean 'O_SYNC'?
  cargo:warning=                        ((fl & O_DSYNC)     ? __WASI_FDFLAG_DSYNC     : 0) |
  cargo:warning=                               ^
  cargo:warning=wasm3/source/m3_api_defs.h:37:58: note: in definition of macro 'm3ApiWriteMem16'
  cargo:warning= #  define m3ApiWriteMem16(ptr, val)  { * (u16 *)(ptr) = (val); }
  cargo:warning=                                                          ^~~
  cargo:warning=wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_fd_seek':
  cargo:warning=wasm3/source/m3_api_wasi.c:353:11: warning: implicit declaration of function 'lseek'; did you mean 'fseek'? [-Wimplicit-function-declaration]
  cargo:warning=     ret = lseek(fd, offset, whence);
  cargo:warning=           ^~~~~
  cargo:warning=           fseek
  cargo:warning=wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_path_open':
  cargo:warning=wasm3/source/m3_api_wasi.c:422:58: error: 'O_DSYNC' undeclared (first use in this function); did you mean 'O_SYNC'?
  cargo:warning=                 ((fs_flags & __WASI_FDFLAG_DSYNC)      ? O_DSYNC     : 0) |
  cargo:warning=                                                          ^~~~~~~
  cargo:warning=                                                          O_SYNC
  cargo:warning=wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_fd_read':
  cargo:warning=wasm3/source/m3_api_wasi.c:474:19: warning: implicit declaration of function 'read'; did you mean 'fread'? [-Wimplicit-function-declaration]
  cargo:warning=         int ret = read (fd, addr, len);
  cargo:warning=                   ^~~~
  cargo:warning=                   fread
  cargo:warning=wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_fd_write':
  cargo:warning=wasm3/source/m3_api_wasi.c:509:19: warning: implicit declaration of function 'write'; did you mean 'fwrite'? [-Wimplicit-function-declaration]
  cargo:warning=         int ret = write (fd, addr, len);
  cargo:warning=                   ^~~~~
  cargo:warning=                   fwrite
  cargo:warning=wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_fd_close':
  cargo:warning=wasm3/source/m3_api_wasi.c:524:15: warning: implicit declaration of function 'close'; did you mean 'pclose'? [-Wimplicit-function-declaration]
  cargo:warning=     int ret = close(fd);
  cargo:warning=               ^~~~~
  cargo:warning=               pclose
  cargo:warning=wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_clock_res_get':
  cargo:warning=wasm3/source/m3_api_wasi.c:596:9: warning: implicit declaration of function 'clock_getres' [-Wimplicit-function-declaration]
  cargo:warning=     if (clock_getres(clk, &tp) != 0) {
  cargo:warning=         ^~~~~~~~~~~~
  cargo:warning=wasm3/source/m3_api_wasi.c: In function 'm3_wasi_unstable_clock_time_get':
  cargo:warning=wasm3/source/m3_api_wasi.c:618:9: warning: implicit declaration of function 'clock_gettime'; did you mean 'localtime'? [-Wimplicit-function-declaration]
  cargo:warning=     if (clock_gettime(clk, &tp) != 0) {
  cargo:warning=         ^~~~~~~~~~~~~
  cargo:warning=         localtime
  exit code: 1

  --- stderr


  error occurred: Command "arm-none-eabi-gcc" "-O0" "-ffunction-sections" "-fdata-sections" "-g" "-fno-omit-frame-pointer" "-mthumb" "-march=armv7e-m" "-I" "wasm3/source" "-Dd_m3LogOutput=0" "-Dd_m3HasWASI" "-Dd_m3Use32BitSlots=1" "-o" "/Users/jiayihu/Desktop/Repo/fedra-thesis/fedra/target/thumbv7em-none-eabi/debug/build/wasm3-sys-9764b8bd4a89ecee/out/wasm3/source/m3_api_wasi.o" "-c" "wasm3/source/m3_api_wasi.c" with args "arm-none-eabi-gcc" did not execute successfully (status code exit code: 1).
```

It seems like I'm unable to compile wasm3/source with `arm-none-eabi-gcc`, do you have any idea?

**Comment by Veykril** [comment]
This might be on me, given that it only seems to fail due to wasi stuff. Can you see if this also happens when building without the wasi feature?
```toml
wasm3 = { version = *, default-features = false, features = ["std", "use-32bit-slots"] }
```

**Comment by Veykril** [comment]
Looks like for some reason it doesn't find this header/the defines in there https://github.com/wasm3/wasm3/blob/a207a3328ff64a3392a575f112582f6d4c12a5d7/source/extra/wasi_core.h

**Comment by jiayihu** [comment]
Can't compile it using `wasm3 = { version = *, default-features = false, features = ["std", "use-32bit-slots"] }` because I'm in `no_std` with `thumbv7em-none-eabi` target. If I just use `features = ["use-32bit-slots"]` the error is much nicer:

```
error: no global memory allocator found but one is required; link to std or add `#[global_allocator]` to a static item that implements the GlobalAlloc trait.

warning: unused import: `rtic::cyccnt::U32Ext`
  --> src/main.rs:12:5
   |
12 | use rtic::cyccnt::U32Ext;
   |     ^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused import: `rprint`
  --> src/main.rs:13:18
   |
13 | use rtt_target::{rprint, rprintln, rtt_init_print};
   |                  ^^^^^^

error: `#[alloc_error_handler]` function required, but not found

error[E0308]: mismatched types
  --> src/main.rs:34:41
   |
34 |         let mut cp: rtic::Peripherals = cx.core;
   |                     -----------------   ^^^^^^^ expected struct `rtic::Peripherals`, found struct `cortex_m::Peripherals`
   |                     |
   |                     expected due to this

error: aborting due to 3 previous errors; 2 warnings emitted

For more information about this error, try `rustc --explain E0308`.
error: could not compile `fedra`.

To learn more, run the command again with --verbose.
```

For some reason the code that was working before including wasm3 now gives error, but I can ignore it for now. The main error is the absense of an allocator, probably because you are assuming std io available right? In no_std I don't have an allocator. I've also noticed that ` BINDGEN_EXTRA_CLANG_ARGS="--sysroot=/usr/local/opt/gcc-arm-none-eabi/arm-none-eabi"` is not needed anymore, so I fear that the previous error is just covered by these simpler one.

**Comment by Veykril** [comment]
The crate still relies on `alloc` atm due to the use of ref counted pointers as this makes the api easier to use. I couldn't come up with an api that didn't rely on that yet. So for some reason enabling wasi doesn't work properly for you here but I'm not sure why to be honest.

**Comment by jiayihu** [comment]
Not a big issue atm, I'd just be able to do 2+2 in WASM 😄 So for now it means that I have to implement an allocator, maybe one from phil's tutorial? https://os.phil-opp.com/allocator-designs/#implementation-for-locked-bumpallocator

**Comment by Veykril** [comment]
Maybe you can use https://github.com/rust-embedded/alloc-cortex-m?

**Comment by jiayihu** [comment]
Alright, forgot about that one thanks! Okay I'll let you know then, thanks very much for the kind support

**Comment by Veykril** [comment]
No problem. Just letting you know that I don't know though whether I can look into the wasi problem atm as its been quite a bit since i properly worked on this crate myself + I only have a windows and a linux machine to test with so I cant even reproduce this problem.

**Comment by Veykril** [comment]
That looks like its missing libc for linking.

**Comment by Veykril** [comment]
This could partially be caused by my buildscript though. It's not the best and it is definitely missing some things.

**Comment by jiayihu** [comment]
Some progress again, although it may not help too much because AFAIK this crate doesn't use cmake. But being unable to compile it with bindgen, I resorted to trying to compile the original wasm3 and after some nasty issues with clang I was able to solve it: https://github.com/wasm3/wasm3/issues/174 

It's also compiled with WASI support, so the issue is something about bindgen or the buildscript, not wasm3 or clang I suppose.

**Comment by Veykril** [comment]
The [buildscript](https://github.com/Veykril/wasm3-rs/blob/master/wasm3-sys/build.rs) is supposed to replace the cmake configuration, but as you might be able to tell as well, it's not complete. So it might be missing a flag required for your platform to work atm.

**Comment by jiayihu** [comment]
#21 solved the issue for me. There is still an open question about how to deal with wasm3 compile flags because compiling the C code for embedded and linking with Rust lld results in [missing intrinsics](https://github.com/rust-lang/compiler-builtins/issues/387), which could be avoided setting `d_m3HasFloat` and `d_m3FixedHeap` if one desire it. But that's another issue, we can close this one
